### PR TITLE
Feature/fridge backend

### PR DESCRIPTION
**Pull request by CapraSnitzel**
### What has been changed?
In this PR, I have implemented methods that allow us to add, remove and update ingredients in the users fridge. These methods are:
 - `removeIngredientFromUserFridge`: method that removes a specific ingredient from the users fridge
 - `updateIngredientFromFridge`: method that adds when the ingredient is not present in the fridge and updates with a new quantity if it is in the fridge; in addition, there is a parameter `scannedItem` which indicates if the method is called for an item that was scanned using the barcode, in which case we can scan multiple times to add multiple times the quantity

In addition, in order to retrieve the correct ingredients from the databse, a `FridgeItem` data class was added to be stored in database:
```kotlin
data class FridgeItem(val id: String, val quantity: String, val expirationDate: LocalDate)
```
This data class contains the id which represents the barcode of the ingredient, the quantity that we actually have and the expiration date of the ingredient. In the database, it looks like this:

image

The expiration date contains more fields than required becase we store a `LocalDate`.

In order to store the fridge items, the way we store the user had to be changes. Thus, we have a new data structure for the user:
```kotlin
data class User(
    val uid: String,
    val userName: String,
    val profilePictureUrl: String,
    val fridge: List<FridgeItem>,
    val likedRecipes: List<String>,
    val createdRecipes: List<String>
)
```
Because of this, we also had to change the way we translate the `fridge` field of the `DocumentSnapshot`.
The choice of storing `List<FridgeItem>` in the database is because we dont want to store twice the ingredients: once in the user and another in the database of the ingredients. Thus, each time we fetch the user, we also fetch the ingredients from the database.
### Why was this change made?
These changes were made in order to be able to use the UI of the fridge which will be implemented in the future and for the user to be able to add his ingredients to the fridge and be able to see when they expire.
## Checklist
- [x] Tests added.
- [ ] Documentation updated.
- [x] All CI checks passed.
- [x] Linked issue/feature (if applicable): #233

### CODE REVIEW

File: app/src/main/java/com/android/sample/model/user/UserViewModel.kt [hunk 1]
```diff
@@ -192,54 +196,94 @@ class UserViewModel(
...
+      expirationDate: LocalDate,
+      scannedItem: Boolean
+  ) {
+    try {
+      val changedIngredient: Pair<FridgeItem, Ingredient> =
+          _fridgeItems.value.first {
+            (it.first.id == ingredient.barCode.toString()) &&
+                (it.first.expirationDate == expirationDate)
+          }
+      if (scannedItem) { // case there we scan an item that we want to add to the fridge
```
Comment (Ylanv) [inline app/src/main/java/com/android/sample/model/user/UserViewModel.kt:254]: I find this idea intriguing but also somewhat concerning in its current form. From what I understand, the goal is to handle cases where a user scans the same item multiple times. For example, if someone buys four strawberry yogurts, the system would create a single FridgeItem and update the quantity accordingly.

While I appreciate the practicality of this approach, I see a potential issue: four individual yogurts of 100g each are not the same as a single large pot of 400g. For instance, if the user eats one yogurt daily, it’s much simpler to track and remove one yogurt at a time rather than adjusting a larger quantity manually.

In conclusion, I believe we need to pay extra attention to defining the unit of an item in the fridge. Should it represent the total weight, or should it reflect the number of individual pieces? Each approach has implications for usability and practicality. What do you think?

File: app/src/main/java/com/android/sample/model/user/UserViewModel.kt [hunk 2]
```diff
@@ -192,54 +196,94 @@ class UserViewModel(
...
-        updateList(_fridge, changedElement, add = false)
-        updateList(_fridge, Pair(changedElement.first, changedElement.second - count), add = true)
-      } else if (changedElement.second - count == 0) {
-        updateList(_fridge, changedElement, add = false)
-        updateList(_fridge, changedElement, add = false)
-      } else {
-        throw IllegalArgumentException(REMOVED_TOO_MANY_INGREDIENTS_ERROR)
-      }
+      val deletedIngredient: Pair<FridgeItem, Ingredient> =
+          _fridgeItems.value.first { it.first.id == ingredient.barCode.toString() }
```
Comment (Ylanv) [inline app/src/main/java/com/android/sample/model/user/UserViewModel.kt:225]: It seems that you are deleting an ingredient by only checking its ID. However, based on the rest of your implementation, it appears that the same ingredient can exist with different expiration dates (which is great, your handling of this in `updateIngredientFromFridge` is a nice feature, allowing users to track multiple entries of the same item with distinct expiration dates).

This approach is problematic because the first ingredient matching the barcode may not necessarily be the one the user intends to delete. To address this, the function should also accept the expiration date as a parameter and use it to identify the correct ingredient to delete. For example, you could incorporate logic similar to what you used in `updateIngredientFromFridge`:

```kotlin
val changedIngredient: Pair<FridgeItem, Ingredient> = 
    _fridgeItems.value.first { 
        (it.first.id == ingredient.barCode.toString()) && 
        (it.first.expirationDate == expirationDate) 
    }
```

This would ensure that the deletion targets the intended ingredient. What do you think?

File: app/src/main/java/com/android/sample/model/user/UserViewModel.kt [hunk 3]
```diff
@@ -192,54 +196,94 @@ class UserViewModel(
...
+      scannedItem: Boolean
+  ) {
+    try {
+      val changedIngredient: Pair<FridgeItem, Ingredient> =
+          _fridgeItems.value.first {
+            (it.first.id == ingredient.barCode.toString()) &&
+                (it.first.expirationDate == expirationDate)
+          }
+      if (scannedItem) { // case there we scan an item that we want to add to the fridge
+        val actualQuantity = regexForQuantity.find(changedIngredient.first.quantity)!!.value
```
Comment (Ylanv) [inline app/src/main/java/com/android/sample/model/user/UserViewModel.kt:255]: The line `val actualQuantity = regexForQuantity.find(changedIngredient.first.quantity)!!.value` could potentially throw a `NullPointerException` that isn’t caught by the current `try-catch` block, which could result in the app crashing. 

Additionally, I’m not entirely comfortable with the way this method is implemented. For instance, the use of `_fridgeItems.value.first` can lead to a `NoSuchElementException` if no element is found, as stated in the [documentation](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/first.html). Instead, using the [find](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/find.html) method would avoid throwing an exception if no element is found, as it would simply return `null`. 

This approach would also allow us to eliminate the `try-catch` block entirely by replacing it with a straightforward null check using an `if`. I believe this would make the method cleaner and more maintainable. What are your thoughts on this?

File: app/src/main/java/com/android/sample/model/user/UserViewModel.kt
Comment (Ylanv) [inline app/src/main/java/com/android/sample/model/user/UserViewModel.kt:1]: Great work on this implementation! To take your PR to the next level, you might consider adding a schema or diagram to describe the overall backend structure for the fridge. Since I was involved in the discussions, I understand why you use a `List<Pair<FridgeItem, Ingredient>>` inside the `ViewModel` while only a `List<FridgeItem>` is stored in Firestore. However, this could be confusing for someone less familiar with the context.

You don’t **have** to include a schema, but adding a brief explanation of this implementation choice could greatly enhance clarity for reviewers and future contributors. It would make your PR even more robust. What do you think?

File: app/src/main/java/com/android/sample/model/user/UserRepositoryFirestore.kt
Comment (Ylanv) [inline app/src/main/java/com/android/sample/model/user/UserRepositoryFirestore.kt:1]: Great job on this implementation! I was curious about how this would look in Firestore, so I checked the database to see if there were any users with a non-empty fridge. **Unfortunately**, I couldn’t find any examples.

It would be a great addition to the PR if you could include a screenshot of the data structure for the `Fridge` field in Firestore. This would really help reviewers understand what’s happening inside `fridgeItemExtraction` and provide more context for your implementation. What do you think?

File: app/src/test/java/com/android/sample/model/user/UserRepositoryFirestoreTest.kt
Comment (Ylanv) [inline app/src/test/java/com/android/sample/model/user/UserRepositoryFirestoreTest.kt:1]: Great work on the Firestore mapping test! Everything looks good to me (LGTM).

File: app/src/test/java/com/android/sample/model/user/UserViewModelTest.kt
Comment (Ylanv) [inline app/src/test/java/com/android/sample/model/user/UserViewModelTest.kt:1]: Overall great testing, very good line coverage and condition coverage. If you want to really push it to the limit i think it would be good to test the same ingredient but with different expiration date which is not tested if im not mistaken.

**Review by Ylanv — CHANGES_REQUESTED** [review]
Great job overall! I’ve suggested some changes that I believe are necessary, along with a few nitpicky details. Feel free to prioritize the necessary changes and address the smaller details only if you have the time.

File: app/src/main/java/com/android/sample/model/user/UserViewModel.kt
Comment (CapraSnitzel) [reply app/src/main/java/com/android/sample/model/user/UserViewModel.kt:1]: Good idea. I did not think about that. I will add it to the PR right away.

File: app/src/main/java/com/android/sample/model/user/UserRepositoryFirestore.kt
Comment (CapraSnitzel) [reply app/src/main/java/com/android/sample/model/user/UserRepositoryFirestore.kt:1]: Same as before, i did not think about adding this to the PR. I will also add it right away.

File: app/src/test/java/com/android/sample/model/user/UserViewModelTest.kt
Comment (CapraSnitzel) [reply app/src/test/java/com/android/sample/model/user/UserViewModelTest.kt:1]: I added couple of tests that adds the same ingredient with different expiration dates as well as deleting one of them and checking is the correct one was deleted.

File: app/src/test/java/com/android/sample/model/user/UserRepositoryFirestoreTest.kt
Comment (CapraSnitzel) [reply app/src/test/java/com/android/sample/model/user/UserRepositoryFirestoreTest.kt:1]: Thanks !

File: app/src/main/java/com/android/sample/model/user/UserViewModel.kt [hunk 4]
```diff
@@ -84,12 +87,13 @@ class UserViewModel(
         onSuccess = { user ->
           _userName.value = user.userName
           _profilePictureUrl.value = user.profilePictureUrl
-          user.fridge.forEach { (barcode, ingredientCount) ->
+          user.fridge.forEach { fridgeItem ->
             ingredientRepository.get(
-                barcode.toLong(),
```
[the remark this answers is not in the document]
Comment (CapraSnitzel) [reply app/src/main/java/com/android/sample/model/user/UserViewModel.kt:89]: I created a private method called `fetchIngredient` that does the following action:
```kotlin
ingredientRepository.get(
        fridgeItem.id.toLong(),
        onSuccess = { ingredient ->
          if (ingredient != null) {
            updateIngredientFromFridge(
                ingredient, fridgeItem.quantity, fridgeItem.expirationDate, false)
          } else {
            Log.e(LOG_TAG, NOT_FOUND_INGREDIENT_IN_DATABASE_ERROR)
          }
        },
        onFailure = { e -> Log.e(LOG_TAG, FAILED_TO_FETCH_INGREDIENT_FROM_DATABASE_ERROR, e) })
````
Let me know what you think.

File: app/src/main/java/com/android/sample/model/user/UserViewModel.kt (hunk 3 above)
Comment (CapraSnitzel) [reply app/src/main/java/com/android/sample/model/user/UserViewModel.kt]: Concerning the `find` method of the regex, it is no longer used. Thus, there is no more concerns about a `NullPointerException`.
About the `_fridgeItems.value.first`, I replaced the `.first` with `.find` like you suggested and replaced the `try-catch` block by an `if-else` statement. Tell me what you think about it.

File: app/src/main/java/com/android/sample/model/user/UserViewModel.kt (hunk 2 above)
Comment (CapraSnitzel) [reply app/src/main/java/com/android/sample/model/user/UserViewModel.kt]: Good catch. I made the decision to add multiple times the same ingredient but with different expiration dates after the implementation of the deletion of an ingretient. Thus I forgot to update the method.
It is updated now.

File: app/src/main/java/com/android/sample/model/user/UserViewModel.kt (hunk 1 above)
Comment (CapraSnitzel) [reply app/src/main/java/com/android/sample/model/user/UserViewModel.kt:241]: After talking with @AdriB02, I have chosen to make the quantity an `Int` in order to represent the number of individual pieces. This has been chosen because it makes the implementation in the UI much easier. Tell me if you are ok with this decision.

File: app/src/main/java/com/android/sample/model/user/UserViewModel.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ylanv) [reply app/src/main/java/com/android/sample/model/user/UserViewModel.kt:89]: Perfect

File: app/src/main/java/com/android/sample/model/user/UserViewModel.kt (hunk 1 above)
Comment (Ylanv) [reply app/src/main/java/com/android/sample/model/user/UserViewModel.kt:241]: Yes!

File: app/src/main/java/com/android/sample/model/user/UserViewModel.kt (hunk 3 above)
Comment (Ylanv) [reply app/src/main/java/com/android/sample/model/user/UserViewModel.kt]: LGTM

File: app/src/main/java/com/android/sample/model/user/UserViewModel.kt
Comment (Ylanv) [reply app/src/main/java/com/android/sample/model/user/UserViewModel.kt:1]: Thanks

**Review by Ylanv — APPROVED** [review]
